node/src/actors/sessions_manager/handlers.rs

```diff
@@ -210,3 +212,39 @@ where
             });
     }
 }
+
+impl Handler<EpochNotification<()>> for SessionsManager {
+    type Result = ();
+
+    fn handle(&mut self, _msg: EpochNotification<()>, ctx: &mut Context<Self>) {
+        let all_ready_before = self.beacons.iter().all(|(_k, v)| v.is_some());
+        if !all_ready_before {
+            // Some peers sent us beacons, but not all of them
+            self.send_peers_beacons(ctx);
+        }
+        // New epoch, new beacons
+        self.clear_beacons();
```

Comment:
There is a race condition here: we must receive the beacons after this handler has been executed. We could avoid this by only clearing beacons from past epochs, and accepting beacons for future epochs in this line:

             // We already got all the beacons for this epoch, do nothing

Reply:
Please write a TODO